### PR TITLE
[Quant] [PT2] Enable batchnorm in _move_exported_model_to_eval

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114547

**Summary**
Add standalone batchnorm into `_move_exported_model_to_eval` to move it from training mode into eval mode

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_bn_conv2d
python -u -m pytest -s -v test_quantize_pt2e.py -k test_bn_move_exported_model_to_eval
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D51853407](https://our.internmc.facebook.com/intern/diff/D51853407)

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

File: torch/ao/quantization/pt2e/eval_utils.py
Comment (andrewor14) [inline torch/ao/quantization/pt2e/eval_utils.py:103]: In L102, can you add batchnorm to the description?

File: test/quantization/pt2e/test_quantize_pt2e_qat.py
Comment (andrewor14) [inline test/quantization/pt2e/test_quantize_pt2e_qat.py:163]: Why this change?

**Review by andrewor14 — COMMENTED** [review]
Hi @leslie-fang-intel, thanks for working on this. Could you also add a check for batchnorm in the [existing test](https://github.com/pytorch/pytorch/blob/9f073ae304263f13168a661a0608149a2eb5b8cc/test/quantization/pt2e/test_quantize_pt2e.py#L1609)?

**Comment by andrewor14** [comment]
Actually, can you paste here what the graphs look like before and after the switch? I would expect to see `aten._native_batch_norm_legit(... train=True)` before, and `aten._native_batch_norm_legit_no_training` after the call to `move_exported_model_to_eval`. Is this what you observe in the tests?

**Comment by leslie-fang-intel** [comment]
> Actually, can you paste here what the graphs look like before and after the switch? I would expect to see `aten._native_batch_norm_legit(... train=True)` before, and `aten._native_batch_norm_legit_no_training` after the call to `move_exported_model_to_eval`. Is this what you observe in the tests?

UT added as `python -u -m pytest -s -v test_quantize_pt2e.py -k test_bn_move_exported_model_to_eval`, take this UT as example:

* Before the graph looks like
```
def forward(self, x):
    arg0, = fx_pytree.tree_flatten_spec(([x], {}), self._in_spec)
    arg0_1 = arg0
    _tensor_constant0 = self._tensor_constant0
    add_ = torch.ops.aten.add_.Tensor(_tensor_constant0, 1);  _tensor_constant0 = None
    empty = torch.ops.aten.empty.memory_format([0], dtype = torch.uint8, layout = torch.strided, device = device(type='cpu'))
    _param_constant0 = self._param_constant0
    _param_constant1 = self._param_constant1
    _tensor_constant1 = self._tensor_constant1
    _tensor_constant2 = self._tensor_constant2
    _native_batch_norm_legit = torch.ops.aten._native_batch_norm_legit.default(arg0_1, _param_constant0, _param_constant1, _tensor_constant1, _tensor_constant2, True, 0.1, 1e-05);  arg0_1 = _param_constant0 = _param_constant1 = _tensor_constant1 = _tensor_constant2 = None
    getitem = _native_batch_norm_legit[0]
    getitem_1 = _native_batch_norm_legit[1]
    getitem_2 = _native_batch_norm_legit[2];  _native_batch_norm_legit = None
    _param_constant2 = self._param_constant2
    _param_constant3 = self._param_constant3
    conv2d = torch.ops.aten.conv2d.default(getitem, _param_constant2, _param_constant3);  getitem = _param_constant2 = _param_constant3 = None
    _tensor_constant3 = self._tensor_constant3
    add__1 = torch.ops.aten.add_.Tensor(_tensor_constant3, 1);  _tensor_constant3 = None
    empty_1 = torch.ops.aten.empty.memory_format([0], dtype = torch.uint8, layout = torch.strided, device = device(type='cpu'))
    _param_constant4 = self._param_constant4
    _param_constant5 = self._param_constant5
    _tensor_constant4 = self._tensor_constant4
    _tensor_constant5 = self._tensor_constant5
    _native_batch_norm_legit_1 = torch.ops.aten._native_batch_norm_legit.default(conv2d, _param_constant4, _param_constant5, _tensor_constant4, _tensor_constant5, True, 0.1, 1e-05);  conv2d = _param_constant4 = _param_constant5 = _tensor_constant4 = _tensor_constant5 = None
    getitem_3 = _native_batch_norm_legit_1[0]
    getitem_4 = _native_batch_norm_legit_1[1]
    getitem_5 = _native_batch_norm_legit_1[2];  _native_batch_norm_legit_1 = None
    return pytree.tree_unflatten([getitem_3], self._out_spec)
```

* After the graph looks like
```
def forward(self, x):
    arg0, = fx_pytree.tree_flatten_spec(([x], {}), self._in_spec)
    arg0_1 = arg0
    _param_constant0 = self._param_constant0
    _param_constant1 = self._param_constant1
    _tensor_constant1 = self._tensor_constant1
    _tensor_constant2 = self._tensor_constant2
    _param_constant2 = self._param_constant2
    _param_constant3 = self._param_constant3
    _native_batch_norm_legit_no_training = torch.ops.aten._native_batch_norm_legit_no_training.default(arg0_1, _param_constant0, _param_constant1, _tensor_constant1, _tensor_constant2, 0.1, 1e-05);  arg0_1 = _param_constant0 = _param_constant1 = _tensor_constant1 = _tensor_constant2 = None
    getitem_6 = _native_batch_norm_legit_no_training[0];  _native_batch_norm_legit_no_training = None
    conv2d = torch.ops.aten.conv2d.default(getitem_6, _param_constant2, _param_constant3);  getitem_6 = _param_constant2 = _param_constant3 = None
    _param_constant4 = self._param_constant4
    _param_constant5 = self._param_constant5
    _tensor_constant4 = self._tensor_constant4
    _tensor_constant5 = self._tensor_constant5
    _native_batch_norm_legit_no_training_1 = torch.ops.aten._native_batch_norm_legit_no_training.default(conv2d, _param_constant4, _param_constant5, _tensor_constant4, _tensor_constant5, 0.1, 1e-05);  conv2d = _param_constant4 = _param_constant5 = _tensor_constant4 = _tensor_constant5 = None
    getitem_7 = _native_batch_norm_legit_no_training_1[0];  _native_batch_norm_legit_no_training_1 = None
    return pytree.tree_unflatten([getitem_7], self._out_spec)

```

File: torch/ao/quantization/pt2e/eval_utils.py
Comment (leslie-fang-intel) [reply torch/ao/quantization/pt2e/eval_utils.py:103]: Thanks and changed.

File: test/quantization/pt2e/test_quantize_pt2e_qat.py
Comment (leslie-fang-intel) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:163]: Previously, `torch.ao.quantization.move_exported_model_to_eval(model_pt2e)` is after `prepare_qat_pt2e` but before `convert_pt2e` in this case. And `convert_pt2e` will do `_fold_conv_bn_qat` which generate `bn` pattern in training mode. So, after we enable `bn` in `move_exported_model_to_eval`, it will fail the conv-bn folding. 
I think we can move it after `convert_pt2e` to make the `_fold_conv_bn_qat` works.

**Comment by leslie-fang-intel** [comment]
Hi @andrewor14, thanks for the comments. I have modified this PR based on it. Please kindly help to take a look of this PR again. Besides, I found `inplace dropout` looks also missing in `_move_exported_model_to_eval`, I am trying to enable it in PR: https://github.com/pytorch/pytorch/pull/114725. Could you also help to take a look?

File: test/quantization/pt2e/test_quantize_pt2e.py
Comment (andrewor14) [inline test/quantization/pt2e/test_quantize_pt2e.py:1647]: Just wanted to understand this better. This test would still pass even if we only have 1 batch norm right? Since we're not doing `_fold_conv_bn_qat` here?

**Review by andrewor14 — APPROVED** [review]
Thanks @leslie-fang-intel. Just left a couple of questions but overall looks good. I'm still not sure if we need to change the order between convert and `move_exported_model_to_eval`, but we can discuss more in the inline comments.

File: test/quantization/pt2e/test_quantize_pt2e_qat.py
Comment (leslie-fang-intel) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:163]: > If that's the case we don't even need move_exported_model_to_eval to support batch norm for the conv-bn case, since there's no batch norm op anymore after folding.

Yes, I think so. If all BN folded into Conv, we don't need to call `move_exported_model_to_eval` which is only needed for single BN case.

> Doesn't the bn pattern disappear after calling _fold_conv_bn_qat

I think here, in the previous code,  `move_exported_model_to_eval` is called before `_fold_conv_bn_qat` which fails this `_fold_conv_bn_qat`.

File: test/quantization/pt2e/test_quantize_pt2e.py
Comment (leslie-fang-intel) [reply test/quantization/pt2e/test_quantize_pt2e.py:1647]: Yes, this UT still pass if we only have 1 BN.

Comment (leslie-fang-intel) [reply test/quantization/pt2e/test_quantize_pt2e.py:1647]: Update the UT to only keep the first BN.

File: torch/ao/quantization/pt2e/eval_utils.py
Comment (leslie-fang-intel) [reply torch/ao/quantization/pt2e/eval_utils.py:62]: Thanks and comment added.

File: test/quantization/pt2e/test_quantize_pt2e_qat.py
Comment (andrewor14) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:163]: Wait, this is actually a problem then. What was the failure? We don't want to impose any ordering requirements between `move_exported_model_to_eval` and `convert_pt2e`. If the test failed in the previous order then we need to fix this

Comment (leslie-fang-intel) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:163]: > What was the failure? 

The checks for `quant_result_pt2e` and `quant_result_fx` will fail in: https://github.com/pytorch/pytorch/blob/235eaabfedbecabd0c5e05ba5a6d92ba578453b1/test/quantization/pt2e/test_quantize_pt2e_qat.py#L171 I think it's due to `_fold_conv_bn_qat_helper` assume BN is in training mode and will fail to do folding after we enabling `_replace_batchnorm_for_eval` in `_move_exported_model_to_eval`.

>  We don't want to impose any ordering requirements between move_exported_model_to_eval and convert_pt2e. If the test failed in the previous order then we need to fix this

ok, then I think may be we can have 2 options:

- Option 1: Enable the Conv-BN folding in `_fold_conv_bn_qat_helper` when BN is in eval mode. Draft a PR for this option: https://github.com/pytorch/pytorch/pull/114927
- Option 2: Makes `_replace_batchnorm_for_eval` optional in `_move_exported_model_to_eval` and makes it default off.

What do you think of it @andrewor14?

**Comment by leslie-fang-intel** [comment]
Hi @andrewor14, could you help to check if it fixes the internal failure?

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51853407).

**Comment by andrewor14** [comment]
Hi @leslie-fang-intel, we're running the tests now. Please don't merge this one yet. I will merge it once all the internal tests pass / once we figure out all the test failures.

**Pull request closed without merging**
